src/eth_rpc/servers/kakarot_rpc.rs

```diff
@@ -0,0 +1,66 @@
+use starknet::core::crypto::compute_hash_on_elements;
+use starknet::core::types::{BroadcastedInvokeTransaction, BroadcastedInvokeTransactionV1};
+use crate::eth_provider::starknet::kakarot_core::to_starknet_transaction;
+use crate::eth_provider::provider::EthereumProvider;
+use starknet::core::types::FieldElement;
+use std::convert::TryInto;
+use jsonrpsee::core::{async_trait, RpcResult};
+use crate::eth_rpc::api::kakarot_api::KakarotApiServer;
+use reth_primitives::B256;
+
+#[derive(Debug)]
+pub struct KakarotRpc<P: EthereumProvider> {
+    eth_provider: P,
+}
+
+impl<P: EthereumProvider> KakarotRpc<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+trait ToElements {
+    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, &'static str>;
+}
+
+impl ToElements for BroadcastedInvokeTransaction {
+    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, &'static str> {
+        match self {
+            BroadcastedInvokeTransaction::V1(tx_v1) => Ok(tx_v1),
+            BroadcastedInvokeTransaction::V3(_) => Err("Transaction is V3, cannot convert to V1"),
+        }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> KakarotApiServer for KakarotRpc<P> {
+    async fn kakarot_get_starknet_transaction_hash(
+        &self,
+        hash: B256,
+        retries: u8,
+    ) -> RpcResult<FieldElement> {
+        // Retrieve the stored transaction from the database.
+        let transaction: reth_rpc_types::Transaction = self.eth_provider.transaction_by_hash(hash).await.unwrap().unwrap();
+
+        // Convert the `Transaction` instance to a `TransactionSigned` instance.
+        let transaction_signed_ec_recovered: reth_primitives::TransactionSignedEcRecovered = <reth_rpc_types::Transaction as TryInto<reth_primitives::TransactionSignedEcRecovered>>::try_into(transaction).unwrap();                
+        let (transaction_signed, _address) = transaction_signed_ec_recovered.to_components();
+
+        // Retrieve the signer of the transaction.
+        let signer = transaction_signed.recover_signer().unwrap();
+
+        // Create the Starknet transaction.
+        let starknet_transaction = (to_starknet_transaction(&transaction_signed, signer, retries).unwrap()).try_into_v1().unwrap();
+
+        let mut data_to_hash: Vec<FieldElement> = vec![];
+        data_to_hash.push(starknet_transaction.sender_address);
+        data_to_hash.extend(starknet_transaction.calldata.clone());
+        data_to_hash.push(starknet_transaction.max_fee);
+        data_to_hash.extend(starknet_transaction.signature.clone());
+        data_to_hash.push(starknet_transaction.nonce);
```

Comment:
you will need to modify this, you can take a look [here](https://docs.starknet.io/architecture-and-concepts/network-architecture/transactions/#v1_deprecated_hash_calculation).

Reply:
Fixed, thanks